### PR TITLE
Major refactor: flatten code and object structures

### PR DESCRIPTION
**Pull request by sempervictus**
Rebuild of PR #31

### CODE REVIEW

**Comment by sempervictus** [comment]
I have no idea what the deuce is up with github these last two days - it somehow managed to offset the master reference. Fixed

File: src/segments.rs [hunk 1]
```diff
@@ -282,10 +286,10 @@ mod tests {
...
             let (f, c, s, oob) = (
-                x.query("F1"),                       //&str
-                x.query("F1.R2"),                    // &str
-                x.query(&*String::from("F1.R2.C1")), //String
-                String::from(x.query("F10")) + x.query("F1.R10") + x.query("F1.R2.C10"),
+                x["F1"],                       //&str
+                x["F1.R2"],                    // &str
+                x["F1.R2.C1"], //String
+                x["F10.R11.C12"]
             );
```
Comment (wokket) [inline src/segments.rs:296]: Previously this test excercised `query(&str)` and `query(String)` to test the `Into<&str>` impl and that has now been removed.   

Can you please re-add the `query` tests outside the string_index tests, and probably update this to test both `[&str]` and `[String]` like the original intent?

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -19,6 +19,6 @@ jobs:
     - name: Build
       run: cargo build --verbose
     - name: Run tests
-      run: cargo test --verbose
+      run: cargo test --verbose --features string_index
     - name: Run Benchmarks
```
Comment (wokket) [inline .github/workflows/ci.yml:23]: Nit: Can we just pass --all-features and not have to touch this again?

File: benches/simple_parse.rs [hunk 3]
```diff
@@ -31,11 +31,14 @@ fn get_msh_and_read_field(c: &mut Criterion) {
...
-            }
+            let msh = m.msh();
+            match msh {
+                Ok(m) => {
+                    let _app =  m.msh_3_sending_application.as_ref().unwrap(); // direct variable access
+                     //println!("{}", _app.value());
+                },
+                Hl7ParseError => panic!("Failed to parse MSH")
+            };
         })
```
Comment (wokket) [inline benches/simple_parse.rs:42]: This benchmark has regressed markedly (from ps to µs scale), is creating the segment twice in the bench etc.  Can  youget this cleaned up and representative please?  The idea of these benches was to compare the relative cost of different ways of retrieving a value.

File: src/fields.rs [hunk 4]
```diff
@@ -20,22 +20,33 @@ impl<'a> Field<'a> {
...
-        let subcomponents = components
+        let components: Vec<&'a str> = input
+            .split(delims.component)
+            .collect::<Vec<&'a str>>()
+            .iter()
+            .map(|c| c.split(delims.repeat).collect::<Vec<&'a str>>())
+            .flatten()
+            .collect();
+        let subcomponents: Vec<Vec<&'a str>> = components
             .iter()
```
Comment (wokket) [inline src/fields.rs:31]: I don't get what's in here, is `components` storing components, or repeat occurances, or both, or ???

File: src/fields.rs [hunk 5]
```diff
@@ -20,22 +20,33 @@ impl<'a> Field<'a> {
...
         Ok(field)
     }
 
+    /// Convenience method to check whether Field is a single data element or broken out into multiple internally
+    // Not used internally - marked to allow dead code as a result
+    #[allow(dead_code)]
+    fn is_subdivided(&self) -> bool {
+        self.components.len() > 1 || self.subcomponents[0].len() > 1
+    }
+
```
Comment (wokket) [inline src/fields.rs:49]: If this isn't used why is it here?

File: src/fields.rs [hunk 6]
```diff
@@ -294,13 +305,37 @@ mod tests {
...
-        assert_eq!(f.query(&*idx0), "yyy&zzz");
-        assert_eq!(f.query("R2.C2"), "zzz");
-        assert_eq!(f.query(oob), "");
+        let f0 = Field::parse_mandatory(Some("xxx^yyy&zzz"), &d).unwrap();
+        let f1 = Field::parse_mandatory(Some("yyy&zzz"), &d).unwrap();
+        let f2 = Field::parse_mandatory(Some("xxx"), &d).unwrap();
+        assert_eq!(f0.is_subdivided(), true);
+        assert_eq!(f1.is_subdivided(), true);
+        assert_eq!(f2.is_subdivided(), false);
+    }
```
Comment (wokket) [inline src/fields.rs:316]: This change has removed tests for calling `query` with different param types and I can't see them moved elsewhere?

File: src/fields.rs (hunk 4 above)
Comment (sempervictus) [reply src/fields.rs:31]: `Components` is storing the _first_ tier of split results - in-field separators and repeats.
`Subcomponents` is storing the splits of the elements in `components`.
If the logic here is incorrect - happy to address that.

File: src/fields.rs [hunk 7]
```diff
@@ -294,13 +305,37 @@ mod tests {
...
+        use super::*;
+        #[test]
+        fn test_string_index() {
+            let d = Separators::default();
+            let f = Field::parse_mandatory(Some("xxx^yyy&zzz"), &d).unwrap();
+            let idx0 = String::from("R2");
+            assert_eq!(f["R2"], "yyy&zzz");
+            assert_eq!(f["R2.C2"], "zzz");
+            assert_eq!(f["R2.C3"], "");
+        }
```
Comment (wokket) [inline src/fields.rs:329]: There's no repeats (or technically a single occurrance) in `xxx^yyy&zzz`, why are we querying with R2 here?

File: src/fields.rs (hunk 5 above)
Comment (sempervictus) [reply src/fields.rs:49]: It isn't used in the library, but as a developer using the library it's often handy to have a convenience method that checks both the subdivision containers for lengths exceeding 1.

File: src/fields.rs (hunk 7 above)
Comment (sempervictus) [reply src/fields.rs:329]: YARGH!!! the git mess is driving me mildly insane. I changed that originally to _have_ a repeated field in it. Fixing

File: .github/workflows/ci.yml (hunk 2 above)
Comment (sempervictus) [reply .github/workflows/ci.yml:23]: I think we need to declare somewhere what `all` would mean, or is that part implicit?

File: src/message.rs [hunk 8]
```diff
@@ -12,61 +12,49 @@ use std::ops::Index;
...
+        let seg = self
             .segments
             .iter()
-            .find_map(|s| match s {
-                segments::Segment::MSH(x) => Some(x),
-                _ => None,
-            })
-            .expect("Failed to find hl7 header");
+            .find(|s| s.fields[0].source == "MSH")
+            .unwrap();
```
Comment (wokket) [inline src/message.rs:39]: Can't we just call `get_segment_by_name("MSH")` here and avoid dupe logic?  Or rely on the spec that `segments[0]` is the MSH and assert that fact (prob faster)?

File: .github/workflows/ci.yml (hunk 2 above)
Comment (wokket) [reply .github/workflows/ci.yml:23]: https://doc.rust-lang.org/cargo/commands/cargo-test.html#feature-selection

File: src/fields.rs (hunk 5 above)
Comment (wokket) [reply src/fields.rs:49]: Then is should be `pub` shouldn't it?

File: src/fields.rs (hunk 7 above)
Comment (sempervictus) [reply src/fields.rs:329]: It does still have the repeating field, sorry, thought git ate stuff.
My understanding of the data structure is that component and repeat separators are the first-tier of subdivision whereas subcomponent separators are the 2nd tier. Should i create a separate vector for split components vs split repeats?

**Comment by wokket** [comment]
OK, I have a feeling that one of us isn't quite on the same page with the repeats thing.  There's [pseudocode in the spec](http://www.hl7.eu/HL7v2x/v251/std251/ch02.html#Heading27) for this, but here's my take on it:

A field has one or more 'occurrence ', separated by the repeat delimiter (default `~`).
An occurrence has one or more `components`, separated by the component delimiter (default `^`).
A component has one or more `sub-components`, separated by the sub component delimiter (default `&`),

So a field of "aaa^bbb" as 1 occurrence (repeat), and two components, each with a single sub component.  You could retrieve "aaa" using a query of "R1.C1", or just "C1" as the R1 is kind of redundant.

A field of "aaa^bbb~ccc^ddd" could also retrieve "aaa" using "R1.C1" or "C1" per above, but would use "R2.C1" to get "ccc".

If we had "aaa^bbb&ccc" then we have one occurrence, 2 components, and the second component has two subcomponents.  We could retrieve "ccc" using "C2.SC2".


Do you know something I don't?

File: benches/simple_parse.rs (hunk 3 above)
Comment (sempervictus) [reply benches/simple_parse.rs:42]: Removed the `seg` big from there, though the benchmark should now take longer as `msh()` calls `MshSegment::parse` every time now that the data structure itself does not hold one internally (instead keeping all elements generic).
If we want it to behave "the same way" as it did before, we'd need to move the MSH parse to outside of the loop

File: src/fields.rs (hunk 6 above)
Comment (sempervictus) [reply src/fields.rs:316]: A bit confused on this one. Retransmit? :)

**Comment by sempervictus** [comment]
Ah! Thank you!
So the logic really should be more along the lines of having 3 vectors per field:
1. Repeats
2. Components of repeats (Vec<Vec>)
3. Subcomponents of components? (Vec<Vec<Vec>>)
I've never seen notation accessing the subcomponent by string query, so i think i screwed up the implementation of field actually.
Is the logic above correct? If so, gimme a few and i'll get it implemented for the PR.

**Comment by wokket** [comment]
That's generally correct, but is also where I headed in the first iteration of this library.  When the majority of fields are only a "simple" field with a single value ("blah") the cost of maintaing multiple vectors (in terms of brain power as much as any runtime cost) gets really high.

That's why I'd started down the enum path of having a "SimpleField", a "RepeatingField" etc etc, but hadn't got as far as a good working model when you came along with the path based stuff which is great, means you don't need to build and then deconstruct a bunch of nested vecs all the time, you can just query in and split as needed based on the query.

As an example, if I have a field with a simple value "blah" (ie one occurrance, one component, one sub-component), do you have a vec[0][0][0] == "blah" ?  or just vec[0] = "blah" ?  How do you then index your field?  does `field[0] == field[0,0,0]` == field[0][0][0]?

This is why I've been pushing down the external `Selecter::query()` line of thinking...

**Comment by sempervictus** [comment]
All makes sense, thank you. Impl mostly done, getting test in line with the new structure.
In terms of cost... i was thinking maybe there's some way to get the compiler to agree to let us replace `Vec<'a str>` types with `&[&'a str]` slices which should be generally cheaper.
Alternatively, instead of enums, we might be able to do this by having a "simple field" initializer which skips all that work and just assigns empty `Vec`'s to the struct members. Lemme wrap up the refactor for field and push to review, and then i'll see what i can do about optimizing away some of the added cost.

**Comment by wokket** [comment]
Yeah let's not make this change any larger than it already is.  The cost for me was mainly brain tax.  

Do I have to specify a repeat[0] if I want a component?  
If a component has subcomponents but I don't specify which one do I get the raw string value?  or an error? or ??
A general rule that you have to include all 'higher' levels if you want a 'lower' value makes sense for indexing, but gets in the way if you want (say) the second subcomponent of the only component, in the only repeat.

When you're doing simple indexing into vecs (or arrays, whatever) those questions are really hard, but the path selection stuff you wrote is a much nicer direction, as _we_ get to describe the semantics of the query.  "SC2" vs field[0][0][1] is much easier on the neurons.

**Comment by sempervictus** [comment]
you do not have to specify `repeat[0]` to get a component, that would just yield the entire field in most cases where repeats are not implemented. Components are now where subcomponents were previously - so `component[0]` would get you the first repeat's worth of components:
```
Field {
    source: "x&x^y&y~a&a^b&b",
    delims: Separators {
        segment: '\r',
        field: '|',
        repeat: '~',
        component: '^',
        subcomponent: '&',
        escape_char: '\\',
    },
    repeats: [
        "x&x^y&y",
        "a&a^b&b",
    ],
    components: [
        [
            "x&x",
            "y&y",
        ],
        [
            "a&a",
            "b&b",
        ],
    ],
    subcomponents: [
        [
            [
                "x",
                "x",
            ],
            [
                "y",
                "y",
            ],
        ],
        [
            [
                "a",
                "a",
            ],
            [
                "b",
                "b",
            ],
        ],
    ],
}
```

File: benches/simple_parse.rs (hunk 3 above)
Comment (wokket) [reply benches/simple_parse.rs:42]: Then move the `msh` out of the loop?  This benchmark was testing the speed of retrieving a local member rather doing a vec lookup or query.  If we want a bench for converting a generic segment into a strongly typed segment then that makes sense, but not here.

File: src/fields.rs (hunk 6 above)
Comment (wokket) [reply src/fields.rs:325]: The [code diff](https://github.com/wokket/rust-hl7/pull/34/files/ba3bab989cd0c51b0f53d3ebbf4ea1bce354cf0c#diff-404f498bc0b18e53fc220c0c82becbb727214348f601ca39c090778009bc41c0L299-R315) shows test calls to `query` being removed...

**Comment by wokket** [comment]
OK I'm AFK for a bit.  Ping me when you've worked through whatever you want to do, had a look at the other points I raised in the PR and run it through a clippy pass, and I'll double check it later, but I'd expect it to be pretty close by then.

**Comment by sempervictus** [comment]
Thank you. Grokking clippy and going over the comments again.

File: src/message.rs (hunk 8 above)
Comment (sempervictus) [reply src/message.rs:38]: Ah, yeah... about that: i've found that method calls in the Indexers are frowned upon by the borrow checker, so have gotten into the habit of using local primitives within the function (and avoiding vecs) to keep it from yelling at me. Will take a pass at DRYing it up a bit.

File: src/fields.rs (hunk 7 above)
Comment (sempervictus) [reply src/fields.rs:363]: Repeats are implemented - marking resolved

File: src/fields.rs (hunk 6 above)
Comment (sempervictus) [reply src/fields.rs:361]: Restored `query()` tests - passing

File: src/fields.rs (hunk 5 above)
Comment (sempervictus) [reply src/fields.rs]: I dropped this with the added depth and complexity - can re-add if you'd like, but needs some more logic now.

File: benches/simple_parse.rs (hunk 3 above)
Comment (sempervictus) [reply benches/simple_parse.rs:36]: Roger, thanks for the clarification - done.

File: .github/workflows/ci.yml (hunk 2 above)
Comment (sempervictus) [reply .github/workflows/ci.yml:23]: Thank you - done.

File: src/segments.rs (hunk 1 above)
Comment (sempervictus) [reply src/segments.rs]: Split-out and re-added `query()` tests from their `index` relatives - all passing, thanks.

**Comment by sempervictus** [comment]
That was educational :), appreciate the references.
I think i broke the benchmark somehow - code looks sane to me, but GH seems to be hanging on the changed bench. Even i'm sp _that_ unlucky as to have CI coincidentally hang on the thing i just changed. Probably something i dont get about how benchmarking works. When you're back, could you please peek at that and kick me in the appropriate direction if you know it?

**Comment by sempervictus** [comment]
On the depth of `Field` structure - trying to make it use borrowed slices a la
```diff
diff --git i/src/fields.rs w/src/fields.rs
index 71e7f7f..441cb10 100644
--- i/src/fields.rs
+++ w/src/fields.rs
@@ -9,7 +9,7 @@ use std::ops::Index;
 pub struct Field<'a> {
     pub source: &'a str,
     pub delims: Separators,
-    pub repeats: Vec<&'a str>,
+    pub repeats: &[&'a str],
     pub components: Vec<Vec<&'a str>>,
     pub subcomponents: Vec<Vec<Vec<&'a str>>>,
 }
@@ -40,7 +40,7 @@ impl<'a> Field<'a> {
         let field = Field {
             source: input,
             delims: *delims,
-            repeats,
+            repeats: &repeats[..]
             components,
             subcomponents,

```
results in some nay-saying by the borrow-checker:
```rust
error[E0515]: cannot return value referencing local variable `repeats`
  --> src/fields.rs:47:9
   |
43 |             repeats: &repeats[..],
   |                       ------- `repeats` is borrowed here
...
47 |         Ok(field)
   |         ^^^^^^^^^ returns a value referencing data owned by the current function

error: aborting due to previous error

For more information about this error, try `rustc --explain E0515`.
error: could not compile `rust-hl7`
```
without the borrow, it has an unknown size issue with this:
```rust
error[E0277]: the size for values of type `[&'a str]` cannot be known at compilation time
  --> src/fields.rs:12:18
   |
12 |     pub repeats: [&'a str],
   |                  ^^^^^^^^^ doesn't have a size known at compile-time
   |
   = help: the trait `Sized` is not implemented for `[&'a str]`
   = note: only the last field of a struct may have a dynamically sized type
   = help: change the field's type to have a statically known size
```
so i think i'll let sleeping dogs lie for the time being on this.

**Comment by sempervictus** [comment]
Fun note from clippy regarding efficiency: the distinct `MshSegment` type is ~40x larger than a generic `Segment`:
```rust
warning: large size difference between variants
 --> src/segments.rs:8:5
  |
8 |     MSH(MshSegment<'a>),
  |     ^^^^^^^^^^^^^^^^^^^ this variant is 1952 bytes
  |
  = note: `#[warn(clippy::large_enum_variant)]` on by default
note: and the second-largest variant is 48 bytes:
 --> src/segments.rs:7:5
  |
7 |     Generic(Segment<'a>),
  |     ^^^^^^^^^^^^^^^^^^^^
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#large_enum_variant
help: consider boxing the large fields to reduce the total size of the enum
  |
8 |     MSH(Box<MshSegment<'a>>),
  |         ^^^^^^^^^^^^^^^^^^^

warning: 1 warning emitted
```
I definitely broke the benchmark :( - even when borrowing `msh` there, its still stuck showing:
```

Benchmarking Read Field from MSH (variable)
Benchmarking Read Field from MSH (variable): Warming up for 3.0000 s
```
or just a blank output if you look at it while its running and stuck there.

**Comment by sempervictus** [comment]
Either that benchmark method won the lottery or something weird is going on inside the iterator:
```
Benchmarking Read Field from MSH (variable)
Benchmarking Read Field from MSH (variable): Warming up for 3.0000 s
Benchmarking Read Field from MSH (variable): Collecting 100 samples in estimated 5.0000 s (18446744074B iterations)
Benchmarking Read Field from MSH (variable): Analyzing
Criterion.rs ERROR: At least one measurement of benchmark Read Field from MSH (variable) took zero time per iteration. This should not be possible. If using iter_custom, please verify that your routine is correctly measured.
```
while it _appears_ (during benchmarking) to take a rather long time.

**Comment by wokket** [comment]
Benchmark is running fine for me locally, so I'm inclined to merge it and see if the zombie army that powers github actions are back on duty.

Anything else before I merge?

**Pull request merged**